### PR TITLE
syncthing: 0.12.19 -> 0.12.21 + nixos module user support

### PR DESCRIPTION
**Pull request by peterhoeg**
# Things done:
- [X] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [X] NixOS
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
# More

This patch only applies against master due to changes merged recently.

We are basically doing a few things in this, which I can of course break up in case people feel that makes sense:
1. Upgrade to 0.12.21
2. Add syncthing-inotify (init at 0.6.8)
3. Changes to the NixOS module to allow specifying:
   1. If syncthing should run as a system service, and
   2. If we should be using inotify
4. A stub package for syncthing 0.12 as the upcoming 0.13 will again change the protocol.
5. User services for both syncthing and syncthing-inotify so they will run for normal users on a multi-user machine.

Defaults have been set, so that behaviour is the same as before, so there should be no surprises.

Cherrypicking along with `7c8c578` makes it run against `nixos-unstable`.

### CODE REVIEW

**Comment by jagajaga** [comment]
cc @grwlf @rnhmjoj @davidak @ctheune

File: nixos/modules/services/networking/syncthing.nix [hunk 1]
```diff
@@ -88,30 +117,53 @@ in
...
-        } // config.networking.proxy.envVars;
+        serviceConfig = service // {
+          User = cfg.user;
+          Group = optionalString (cfg.user == defaultUser) defaultUser;
+          PermissionsStartOnly = true;
+          ExecStart = "${pkgs.syncthing}/bin/syncthing -no-browser -home=${cfg.dataDir}";
+        };
+      };
 
+      syncthing-inotify = mkIf.useInotify header // {
```
Comment (rnhmjoj) [inline nixos/modules/services/networking/syncthing.nix]: There's a typo here: should be `mkIf cfg.useInotify header // ...`

Comment (rnhmjoj) [inline nixos/modules/services/networking/syncthing.nix]: Also need to enclose it with parenthesis: `(header // { ... });`

File: nixos/modules/services/networking/syncthing.nix [hunk 2]
```diff
@@ -88,30 +117,53 @@ in
...
+        description = "Monitor Syncthing using inotify";
+        after = [ "syncthing.service" ];
+        requires = [ "syncthing.service" ];
+        wantedBy = [ "multi-user.target" ];
         serviceConfig = {
-          User  = cfg.user;
+          User = cfg.user;
           Group = optionalString (cfg.user == defaultUser) defaultUser;
           PermissionsStartOnly = true;
+          ExecStart = "${pkgs.syncthing-inotify}/bin/syncthing -logflags=0";
```
Comment (rnhmjoj) [inline nixos/modules/services/networking/syncthing.nix]: The binary name should be `syncthing-inotify`

**Comment by rnhmjoj** [comment]
syncthing-inotify runs but can't communicate with syncthing (the log says it may be missing the API key). Is this supposed to work out of the box or should be configured manually?

**Comment by peterhoeg** [comment]
It ran perfectly for me when testing it, so I think I may have pushed the wrong commit. Don't waste any more time on this - I'll check it tomorrow and push the proper commit/branch. I'm sorry you wasted your time on this.

**Comment by rnhmjoj** [comment]
Ok, no problem.

**Comment by rnhmjoj** [comment]
Updates?

File: nixos/modules/services/networking/syncthing.nix [hunk 3]
```diff
@@ -58,16 +91,12 @@ in
       package = mkOption {
         type = types.package;
         default = pkgs.syncthing;
-        defaultText = "pkgs.syncthing";
```
Comment (peti) [inline nixos/modules/services/networking/syncthing.nix]: Why did you remove that attribute?

File: nixos/modules/services/networking/syncthing.nix [hunk 4]
```diff
@@ -27,12 +42,30 @@ in
...
         description = ''
-          Syncthing will be run under this user (user must exist,
-          this can be your user name).
+          Syncthing will be run under this user (user will be created if it doesn't exist.
+          This can be your user name).
+        '';
+      };
+
+      group = mkOption {
+        default = defaultUser;
```
Comment (peti) [inline nixos/modules/services/networking/syncthing.nix]: Re-using the `defaultUser` name for the group here doesn't feel right. It's a *user*, not a group.

Besides, why does the service even need its own group? Why not use `nogroup`?

File: nixos/modules/services/networking/syncthing.nix [hunk 5]
```diff
@@ -7,6 +7,21 @@ let
   cfg = config.services.syncthing;
   defaultUser = "syncthing";
```
Comment (peti) [inline nixos/modules/services/networking/syncthing.nix]: It's customary for those user- and group names to have an entry in `nixos/modules/misc/ids.nix`.

File: nixos/modules/services/networking/syncthing.nix [hunk 6]
```diff
@@ -88,30 +117,53 @@ in
...
+        description = "Monitor Syncthing using inotify";
+        after = [ "syncthing.service" ];
+        requires = [ "syncthing.service" ];
+        wantedBy = [ "multi-user.target" ];
         serviceConfig = {
-          User  = cfg.user;
+          User = cfg.user;
           Group = optionalString (cfg.user == defaultUser) defaultUser;
           PermissionsStartOnly = true;
+          ExecStart = "${pkgs.syncthing-inotify}/bin/syncthing-inotify -logflags=0";
```
Comment (peti) [inline nixos/modules/services/networking/syncthing.nix]: It seems to me like the hard-coded use of `pkgs.syncthing-inotify` here completely breaks the abstraction of `cfg.package`. If I'd define `package` to be a variant of `syncthing` (with inotify support enabled), then I'd expect `syncthing-inotify.service` to use *that* package, not `pkgs.syncthing-inotify`.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -13848,6 +13848,8 @@ in
 
   syncthing = go15Packages.syncthing.bin // { outputs = [ "bin" ]; };
   syncthing011 = go15Packages.syncthing011.bin // { outputs = [ "bin" ]; };
+  syncthing012 = go15Packages.syncthing012.bin // { outputs = [ "bin" ]; };
+  syncthing-inotify = go15Packages.syncthing-inotify.bin // { outputs = [ "bin" ]; };
```
Comment (peti) [inline pkgs/top-level/all-packages.nix]: Why is inotify support not enabled by default?

File: nixos/modules/services/networking/syncthing.nix (hunk 5 above)
Comment (rnhmjoj) [inline nixos/modules/services/networking/syncthing.nix]: I made this change: I did set and id for syncthing but also set a username. Is this superfluous?

Comment (peti) [inline nixos/modules/services/networking/syncthing.nix]: Ah, so there *is* an entry for syncthing in `nixos/modules/misc/ids.nix`. Very cool, thanks for clarifying.

**Comment by peterhoeg** [comment]
I am breaking this PR into multiple, separate PRs so we can deal with them one at a time.

**Comment by peterhoeg** [comment]
@peti you raised the issue of making a derivation with support for inotify. What would be the proper way to achieve that instead of how my PR had it as a separate package?

**Comment by peti** [comment]
@peterhoeg, nevermind, I misunderstood the nature of `synthing-inotify`. I hadn't realized that it's a separate package entirely.
